server/mqtt.go

```diff
@@ -4241,52 +4266,93 @@ func (s *Server) mqttProcessPubRel(c *client, pi uint16, trace bool) error {
 // Invoked from the MQTT publisher's readLoop. No client lock is held on entry.
 func (c *client) mqttHandlePubRetain() {
 	pp := c.mqtt.pp
-	if !mqttIsRetained(pp.flags) {
+	isRetained := mqttIsRetained(pp.flags)
+	isSparkbBirth, typ, groupID, nodeID, deviceID := sparkbIsBirth(pp.subject, sparkbNamespacePrefix)
+	if !isRetained && !isSparkbBirth {
 		return
 	}
-	key := string(pp.subject)
+
+	// Always clear the retain flag to deliver a normal published message.
+	defer func() {
+		pp.flags &= ^mqttPubFlagRetain
+	}()
+
 	asm := c.mqtt.asm
-	// Spec [MQTT-3.3.1-11]. Payload of size 0 removes the retained message,
-	// but should still be delivered as a normal message.
-	if pp.sz == 0 {
+	rm := &mqttRetainedMsg{
+		Origin: asm.jsa.id,
+		Msg:    pp.msg, // will copy these bytes later as we process rm.
+		Flags:  pp.flags,
```

Comment:
For a sparkb message, the pp.flags will indicate that the message is not retained, correct? Should it be switched to retained since we are storing it at such?